### PR TITLE
[IDSEQ-2221] Generate presigned s3 url for downloads

### PR DESCRIPTION
**Pull request by j-x-han**
# Description

See IDSEQ-2221. Replace sample download links with presigned S3 urls

# Notes

I found 3 download types (non-host reads, non-host contigs, unmapped reads) using the previous bad pattern of copying S3 files to rails prior to serving them. Let me know if there are any I missed!

image

# Tests

* Verify that the download links in the sample report sidebar download the expected file.
* Verified that attempting to download a file that doesn't exist doesn't result in the page crashing.

### CODE REVIEW

File: app/controllers/samples_controller.rb [hunk 1]
```diff
@@ -991,8 +991,13 @@ def contigs_fasta
...
 
     if contigs_fasta_s3_path
-      @contigs_fasta = get_s3_file(contigs_fasta_s3_path)
-      send_data @contigs_fasta, filename: @sample.name + '_contigs.fasta'
+      filename = @sample.name + '_contigs.fasta'
+      @contigs_fasta_url = get_presigned_s3_url(contigs_fasta_s3_path, filename)
+      if @contigs_fasta_url
+        redirect_to @contigs_fasta_url
+      else
+        send_data nil, filename: filename
```
Comment (lvreynoso) [inline app/controllers/samples_controller.rb:999]: If `get_presigned_s3_url()` returns `nil`, the user will get a 0 byte file download, is that correct? Would it be better to render an error instead?

Comment (gregdingle) [reply app/controllers/samples_controller.rb:999]: at least we should log the error

File: app/helpers/pipeline_outputs_helper.rb [hunk 2]
```diff
@@ -189,6 +189,26 @@ def get_s3_file(s3_path)
     end
   end
 
+  def get_presigned_s3_url(s3_path, filename)
+    s3 = Aws::S3::Resource.new(client: Client)
+    uri_parts = s3_path.split("/", 4)
```
Comment (cdebourcy) [inline app/helpers/pipeline_outputs_helper.rb:194]: I saw the code that parses the URIs repeated in https://github.com/chanzuckerberg/idseq-web/blob/e70919eec874eecfdaa6d324959203279be7817d/app/controllers/samples_controller.rb#L1012 -- worth factoring it out into one of the helpers?

File: app/controllers/samples_controller.rb (hunk 1 above)
Comment (j-x-han) [reply app/controllers/samples_controller.rb]: Good call. Changed to render and log an error instead.

File: app/helpers/pipeline_outputs_helper.rb (hunk 2 above)
Comment (j-x-han) [reply app/helpers/pipeline_outputs_helper.rb]: Done! I found and replaced some other places where I saw this repeated.

File: app/controllers/samples_controller.rb [hunk 3]
```diff
@@ -1010,9 +1019,7 @@ def contigs_summary
 
   # TODO(mark): Factor out into S3Helper file.
```
Comment (gregdingle) [inline app/controllers/samples_controller.rb:1020]: might as well change this TODO

File: app/helpers/s3_helper.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+module S3Helper
```
Comment (gregdingle) [inline app/helpers/s3_helper.rb:1]: I know you are responding to Mark's comment but I'd rather see this in `S3Util`. I think it is better to use helpers only when sharing instance state. Then the dependencies are clearer.

Comment (j-x-han) [reply app/helpers/s3_helper.rb]: Sounds good, I've moved it to `S3Util`!

**Pull request merged**